compiler/aarch64/codegen/OMRCodeGenerator.hpp

```diff
@@ -286,6 +286,18 @@ class OMR_EXTENSIBLE CodeGenerator : public OMR::CodeGenerator
    TR_GlobalRegisterNumber _gprLinkageGlobalRegisterNumbers[TR::RealRegister::NumRegisters]; // could be smaller
    TR_GlobalRegisterNumber _fprLinkageGlobalRegisterNumbers[TR::RealRegister::NumRegisters]; // could be smaller
 
+   /**
+    * @brief Answers whether a trampoline is required for a direct call to reach a
+    *        target address from some source address.
+    *
+    * @param[in] targetAddress : the absolute address of the call target
+    * @param[in] sourceAddress : the absolute address from which the call displacement
+    *              range is measured.
+    *
+    * @return : true if a trampoline is required; false otherwise.
+    */
+   bool directCallRequiresTrampoline(intptrj_t targetAddress, intptrj_t sourceAddress);
```

Comment:
It does make sense to have a parameter with a clearly defined definition.  The CodeGenerator implementation on each platform should be able to determine what adjustments are needed.  In theory, all platforms except x86 should be passing in the instruction address already based on my understanding of how the displacements are calculated for their respective branch instructions.  X86 will be require a bit of rework that may reach into downstream projects.  I'll look into this deeper and report back.

Reply:
I've made changes to the PR where the sourceAddress of the `directCallRequiresTrampoline` is the address of the call instruction.  For the most part, no changes were required except on x86, and even in that case the changes were minimal.  I'll push those changes shortly.

@fjeremic : in the process of doing this I discovered a couple of places in the Z codegen where I believe the wrong source address is being used already to calculate whether a trampoline is needed.  Here are the two places that I found in the current code base:

https://github.com/eclipse/omr/blob/7a0ff628cb5c514ebaa88c1bb3d884fd8b4ca6fb/compiler/z/codegen/OMRSnippet.cpp#L142

https://github.com/eclipse/omr/blob/7a0ff628cb5c514ebaa88c1bb3d884fd8b4ca6fb/compiler/z/codegen/S390HelperCallSnippet.cpp#L113

In both cases, I believe `cursor` is two bytes off because it has already been advanced once the instruction opcode was written.  If you confirm this is an issue then I'll fix up as part of this work.

Reply:
I should point out that in both of the Z cases above the `cursor` was correctly adjusted during the actual encoding of the instruction (it was just the trampoline test where it is different).